### PR TITLE
Xtheme editor

### PR DESCRIPTION
**Pull request by akx**
So here's an editor for #84 :)

The editor UI is ugly, yes, but I assume @juhakujala will fix it all up...

Ps.: There's a partially pluginified version of the current default theme available here: https://github.com/akx/shoop/tree/xtheme-test-theme

### CODE REVIEW

**Comment by suutari-ai** [comment]
The build fails with:
```
running build_production_resources
    
    > shoop-xtheme@1.0.0 build /tmp/pip-BoEUaW-build/shoop/xtheme
    > gulp --production
    
    sh: 1: gulp: not found
```
You're calling gulp differently here?

**Comment by akx** [comment]
@suutari-ai: it was missing an `npm i &&`, silly me :)

**Comment by suutari-ai** [comment]
Missing migrations?

**Comment by akx** [comment]
... Yes, @suutari-ai. :crying_cat_face: 

I'll fix that tomorrow.

**Comment by Pikkupomo** [comment]
There is a bug in publishing changes:

Browser chrome

* Add new plugin row
* title aaaa
* plugin Product highlights
* count 99999
* save

publish changes
The published changes wont appear in template

To further break:
* open edit panel
* Add new plugin row (aaaa should not be visible)
* title bbbb
* type product highlights
* count 9999
* save
* publish changes

Now the bbbb is visible on template
=> Exit edit
Now the aaaa is visible on template
=> Edit
The edit panel wont show up, the bbbb comes visible

**Comment by Pikkupomo** [comment]
also template broken in edit mode in product highlights

image

**Comment by akx** [comment]
Good findings, @Pikkupomo :) I'm currently fixing these up.

**Comment by akx** [comment]
Btw,

> also template broken in edit (and editless) mode in product highlights

That's the template that's not officially part of this PR, so...

**Comment by akx** [comment]
@Pikkupomo: bugs fixed, admin module added :)

File: shoop/xtheme/models.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        if not model:  # Nothing loaded? Put ourselves in draft mode.
+            model = SavedViewConfig(status=SavedViewConfigStatus.CURRENT_DRAFT, **svc_kwargs)
+        return model
+
+
+class SavedViewConfigStatus(Enum):
+    CURRENT_DRAFT = 1
+    OLD_DRAFT = 2
+    PUBLIC = 3
```
Comment (Pikkupomo) [inline shoop/xtheme/models.py]: `PUBLIC` `DRAFT` `OLD_VERSION`

File: shoop/xtheme/static_src/injection/index.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        return;
+    }
+    const defaultConfigElement = domElement.querySelector("script.xt-ph-default-layout");
+    const defaultConfigJSON = (defaultConfigElement ? defaultConfigElement.innerHTML : null);
+    const viewName = window.XthemeEditorConfig.viewName;
+    const urlParams = {
+        view: viewName,
+        theme: window.XthemeEditorConfig.themeIdentifier,
+        ph: placeholderName,
+        // TODO: Hopefully we won't get any problems with too-long query strings (2048 is the maximum for IE):
```
Comment (akx) [inline shoop/xtheme/static_src/injection/index.js]: FWIW:

This randomly-generated layout serializes to about 2048 bytes of JSON.

```python
{'name': 'test',
 'rows': [{'cells': [{'config': {'text': 'tpziuwxpgz'},
                      'plugin': 'text',
                      'sizes': {}},
                     {'config': {}, 'plugin': None, 'sizes': {'md': 8}}]},
          {'cells': [{'config': {'text': 'kyhnesnqvx'},
                      'plugin': 'text',
                      'sizes': {}}]},
          {'cells': [{'config': {}, 'plugin': None, 'sizes': {'md': 12}},
                     {'config': {}, 'plugin': None, 'sizes': {'md': 2}}]},
          {'cells': [{'config': {'text': 'ohobnujhtk'},
                      'plugin': 'text',
                      'sizes': {'md': 2}},
                     {'config': {'text': 'coujlplkby'},
                      'plugin': 'text',
                      'sizes': {'md': 1}}]},
          {'cells': [{'config': {'text': 'veuvjhvohj'},
                      'plugin': 'text',
                      'sizes': {}}]},
          {'cells': [{'config': {'text': 'dgvsyseexs'},
                      'plugin': 'text',
                      'sizes': {'md': 6}},
                     {'config': {}, 'plugin': None, 'sizes': {'md': 5}},
                     {'config': {}, 'plugin': None, 'sizes': {'md': 12}},
                     {'config': {'text': 'jpfocvwnui'},
                      'plugin': 'text',
                      'sizes': {'md': 8}},
                     {'config': {'text': 'mxqvkkycid'},
                      'plugin': 'text',
                      'sizes': {'md': 6}}]},
          {'cells': [{'config': {'text': 'sgjdakqwsn'},
                      'plugin': 'text',
                      'sizes': {}},
                     {'config': {}, 'plugin': None, 'sizes': {'md': 11}},
                     {'config': {'text': 'zeprgotptr'},
                      'plugin': 'text',
                      'sizes': {'md': 8}}]},
          {'cells': [{'config': {'text': 'itatblnlaz'},
                      'plugin': 'text',
                      'sizes': {}},
                     {'config': {'text': 'yqqoclpsyl'},
                      'plugin': 'text',
                      'sizes': {'md': 12}}]},
          {'cells': [{'config': {'text': 'cqgibkyuxl'},
                      'plugin': 'text',
                      'sizes': {}},
                     {'config': {}, 'plugin': None, 'sizes': {'md': 6}},
                     {'config': {}, 'plugin': None, 'sizes': {'md': 11}},
                     {'config': {'text': 'mycesgegad'},
                      'plugin': 'text',
                      'sizes': {'md': 12}},
                     {'config': {'text': 'owbewefrfg'},
                      'plugin': 'text',
                      'sizes': {'md': 12}},
                     {'config': {}, 'plugin': None, 'sizes': {'md': 3}},
                     {'config': {'text': 'cplnacufdu'},
                      'plugin': 'text',
                      'sizes': {'md': 3}}]},
          {'cells': [{'config': {}, 'plugin': None, 'sizes': {'md': 7}},
                     {'config': {'text': 'hpufawbztn'},
                      'plugin': 'text',
                      'sizes': {'md': 8}},
                     {'config': {'text': 'uupmvloebd'},
                      'plugin': 'text',
                      'sizes': {'md': 12}},
                     {'config': {'text': 'jydcjqvpev'},
                      'plugin': 'text',
                      'sizes': {'md': 6}}]}]}
```

File: shoop/xtheme/theme.py [hunk 3]
```diff
@@ -6,22 +6,137 @@
...
+        :param settings_obj: A theme settings object for this theme, if one exists.
+                             Passing this in will avoid extraneous database queries.
+        :type settings_obj: ThemeSettings|None
+        """
+        self._settings_obj = None
+        if settings_obj and settings_obj.theme_identifier == self.identifier:  # fine, let's accept that
+            self._settings_obj = settings_obj
+
+    @property
+    def settings_obj(self):
```
Comment (Pikkupomo) [inline shoop/xtheme/theme.py]: I would like `def settings(self)` even though it's different from django `settings`

File: shoop/xtheme/views/forms.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        """
+        sizes = ["sm", "md"]  # TODO: Parametrize? Currently Bootstrap dependent.
+        sizes.extend(set(self.layout_cell.sizes) - set(sizes))
+        self.sizes = sizes
+        for size in self.sizes:
+            self.fields["size_%s" % size] = forms.IntegerField(
+                label=size.upper(),
+                required=False,
+                min_value=0,
+                max_value=12,  # TODO: Parametrize? Currently Bootstrap dependent.
```
Comment (Pikkupomo) [inline shoop/xtheme/views/forms.py]: this should be documented in requirements for plugin creators.

**Pull request merged**